[2단계 - 출석 다시 구현하기] 제프(변우영) 미션 제출합니다.

horizonpioneer opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs/3949

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
출력을 위해 getter를 사용했고, 해당 데이터를 밖에서 가공했습니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->
안녕하세요 이프!
step 2를 진행하면서 고민했던 부분에 대해 질문 드리고 싶습니다.

### 1. 검증의 책임만 가진 객체도 존재 가치가 있을까요?

닉네임 검증 책임에 대해 고민해봤습니다. 단순히 닉네임 형식만 검증하는 경우라면 닉네임 객체에서 검증하는 것이 자연스럽지만, 이번 미션에서는 출석 기록을 파일에서 불러와 파일에 등록된 닉네임이 맞는지 확인하는 과정이 필요합니다.

이 때문에 검증 책임을 `Attendances` 객체가 가져야 하는지 고민했습니다. 현재 `AttendancesLoader`를 통해 출석부 데이터를 불러와 `Attendances`로 반환하고, `Attendances`에서 출석부의 key 값에 특정 닉네임이 포함되는지를 검증하는 방식으로 구현했습니다.

또 다른 이유는 `AttendancesLoader`에서 파일을 한 줄씩 읽어 닉네임과 출석 날짜를 파싱하는 과정에서, 닉네임만 따로 모으는 메서드를 추가하면 파일을 두 번 읽어야 하므로 비효율적이라고 판단했기 때문입니다. 이는 기존 설계와 다르게 `Crew` 객체를 두지 않고 출석 기록을  `Attendances`에서 `Map`의 value로 다루는 방식에서 비롯된 문제라고 생각됩니다.

이러한 경우 닉네임 검증을 담당하는 별도의 객체를 두는 것이 더 나은 접근 방식이 될 수 있을까요?

### 2. Enum 내부 정적 메서드를 도메인 로직에서 사용하는 것이 적절할까요?
정적 메서드는 객체지향보다는 절차지향적인 방식이며, 메모리 효율과 관련해서도 단점이 있다고 알고 있습니다. 하지만 실질적으로 큰 단점으로 느껴지지는 않아 고민이 되었습니다.

현재 enum 내부의 상수를 활용해 특정 상태를 판단하는 용도로 정적 메서드를 사용하고 있는데, 개인적으론 사용 이유로 충분하다고 생각합니다. 하지만 도메인 내부에서 enum의 정적 메서드를 직접 사용하는 것이 좋은 설계인지 고민이 됩니다. 이에 대한 의견을 듣고 싶습니다.

### 3. LocalDateTime을 필드로 두는 객체에서 getter를 구현하는 방식이 고민됩니다.

출석 날짜와 시간을 함께 저장할 수 있는 방식이라 `LocalDateTime`을 필드로 사용했습니다. 하지만 `LocalDate`와 `LocalTime`이 합쳐진 형태다 보니 데이터를 가공할 때 코드가 지저분해 보인다는 느낌을 받았습니다.

특히 출력 시 데이터 가공이 많아 `getLocalDate()`와 `getLocalTime()`을 별도로 제공하는 것이 좋을지, 아니면 `getLocalDateTime()`을 그대로 노출한 뒤 외부에서 가공하는 것이 좋을지 고민됩니다.

더 나아가, 필드를 `LocalDateTime`으로 두는 것이 나을지, 아니면 아예 `LocalDate`와 `LocalTime`을 분리하는 것이 나을지도 궁금합니다. 상황에 따라 두 방식 모두 사용할 수 있을 것 같지만, 이렇게 하나의 데이터로 관리할 수 있는 필드를 굳이 분리하는 것이 적절한지 조언 부탁드립니다.

### 4. TDD에서 실패 코드 작성이 오래 걸렸습니다.

TDD 방식으로 구현할 때, 실패 코드를 작성하는 과정에서 시간이 오래 걸렸습니다. 실패 코드를 빠르게 작성하고, 이후 Green 및 Refactoring 단계로 넘어가는 것이 TDD의 핵심이라고 알고 있지만, 실패 방식에 대한 고민이 많았습니다.

이번에는 도메인에 존재하지 않는 기능을 먼저 테스트 코드에서 가정한 뒤, 해당 기능을 도메인에 추가하고 테스트를 실행한 후 리팩토링하는 방식으로 진행했습니다.

이 접근 방식이 올바른지 궁금하며, 이프가 TDD를 시작할 때는 어떤 방식으로 접근하는지 조언 부탁드립니다.

horizonpioneer committed:
step2 initial setting

horizonpioneer committed:
feat: AttendanceStatus enum 추가

horizonpioneer committed:
feat: 출석 시간과 지각 기준 시간, 결석 기준 시간을 바탕으로 출석 상태를 판단하는 기능 추가

horizonpioneer committed:
test: 월요일 출석 상태별 테스트

horizonpioneer committed:
test: 화요일 출석 상태별 테스트

- 월요일 기준 시간을 제외하면 화~금까지 기준 시간이 같으므로, 나머지 요일 테스트를 화요일 테스트로 대체한다.

horizonpioneer committed:
feat: 요일에 따라 기준 시간 enum을 반환하는 기능 추가

horizonpioneer committed:
refactor: judge메서드의 인자를 LocalTime에서 LocalDate로 변경

- 특정 일자에 따라 출석 상태를 판단하도록 수정
- 출석 상태 판단 테스트 코드 수정

horizonpioneer committed:
test: 주말의 경우, 출석 상태 판단 시 예외를 발생시키는 테스트 추가

horizonpioneer committed:
docs: 기능 구현 목록 추가

horizonpioneer committed:
feat: Attendance 추가 및 테스트 코드 추가

- 출석시간을 통해 출석을 체크하는 테스트 구현
- 주말 출석 시, 예외를 발생시키는 테스트 구현

horizonpioneer committed:
feat: 출석 수정 기능 추가 및 테스트 코드 추가

horizonpioneer committed:
refactor: 출석 상태 이름 부여

horizonpioneer committed:
refactor: 출석 상태 판단 로직을 Attendance의 책임으로 수정

horizonpioneer committed:
feat: 크루의 출석 상태별 횟수 조회 기능 추가

- Attendances 클래스 추가
- 크루의 출석 상태별 횟수 조회 테스트 추가

horizonpioneer committed:
style: 오타 수정

horizonpioneer committed:
feat: PenaltyStatus enum 추가 및 제적 대상자 확인 테스트 추가

horizonpioneer committed:
feat: AttendancesLoader 추가 및 테스트 코드 추가

- csv 파일로부터 출석 기록을 불러오는 기능 추가
- 출석 기록을 크루의 출석 목록에 추가하는 기능 추가
- AttendancesLoader 기능에 대한 테스트 코드 추가

horizonpioneer committed:
refactor: 출석 상태 테스트를 AttendanceStatusTest로 이동

horizonpioneer committed:
feat: 출석 기록 csv 파일 추가

horizonpioneer committed:
refactor: 출석 상태 확인 메서드를 AttendanceStatus enum의 정적 메서드로 변경

horizonpioneer committed:
feat: 닉네임과 출석일자를 통해 출석 기록을 가져오는 테스트 추가

horizonpioneer committed:
feat: 닉네임과 수정일자를 통해 출석 기록을 수정하는 테스트 추가

horizonpioneer committed:
feat: 닉네임과 일자를 통해 출석을 추가하는 테스트 추가

horizonpioneer committed:
feat: 해당 크루의 제적 상태를 확인하는 테스트 추가

horizonpioneer committed:
feat: 해당 날짜가 공휴일인지 확인하는 테스트 추가

horizonpioneer committed:
feat: 닉네임과 출석 날짜를 통해 크루의 출석 정보를 가져오는 기능 추가

horizonpioneer committed:
feat: 닉네임과 수정일자를 통해 출석을 수정하는 기능 추가

horizonpioneer committed:
feat: 출석부에 있는 크루의 이름을 모두 반환하는 기능 추가

horizonpioneer committed:
feat: 공휴일에 출석 시 예외를 발생시키는 테스트 추가

horizonpioneer committed:
feat: 커맨드 입력 테스트 추가

horizonpioneer committed:
refactor: PenaltyStatus enum 및 테스트 수정

- 일부 상수를 enum의 필드로 정리
- 패널티 상태 확인 메서드를 stream 로직으로 수정
- 각 패널티 상태별 테스트 추가

horizonpioneer committed:
refactor: PenaltyStatus 내부 메서드 수정

horizonpioneer committed:
refactor: AttendancesTest 수정

horizonpioneer committed:
feat: StandardDate 상수 클래스 추가

horizonpioneer committed:
feat: 출석 csv 파일 수정

horizonpioneer committed:
refactor: 출석 테스트 내부 날짜 수정

horizonpioneer committed:
feat: OutputView 추가

horizonpioneer committed:
feat: AttendanceController 추가

horizonpioneer committed:
feat: Application 추가

horizonpioneer committed:
feat: 컨트롤러 매직넘버 제거

horizonpioneer committed:
feat: 출석부에 존재하지 않는 기록을 결석처리 하는 기능 구현

horizonpioneer committed:
feat: 등록되지 않는 닉네임의 경우 예외 발생 기능 추가

horizonpioneer committed:
refactor: AttendancesLoader 내부 로직 수정

horizonpioneer committed:
refactor: 출석 기록 csv 파일 수정

horizonpioneer committed:
refactor: AttendancesLoader 예외 처리 추가

sinb57 commented on src/main/java/domain/constant/StandardDate.java:
```diff
@@ -1,13 +1,24 @@
 
-    public static final Day TODAY = new Day(LocalDate.of(2024, 12, 26));
+    public static final LocalDate FIXED_DATE = LocalDate.of(2025, 2, 28);
+    public static final LocalDate TODAY = determineToday();
```
오늘의 날짜를 고정하여 도메인에서도 의존하고 있네요. 이렇게 하지 말고, Controller에서 오늘의 날짜를 주입하는 방식으로 로직을 개선해보세요.

horizonpioneer replied:
타임머신을 도입해서 프로그램 실행 시, 오늘 일자를 입력받도록 수정했습니다!

sinb57 replied:
시간을 제어할 수 있도록 하셨군요! 저는 LocalDateTime.now()를 Controller부터 내부 도메인까지 전달하는 구조를 생각했어요ㅎ 지금은 도메인에 대한 테스트를 작성했고 시간에 의존하는 부분이 없기 때문에 필요성을 못 느끼겠지만, 컨트롤러에 대한 테스트를 작성하게 되면 어떨지 고민해보셔요~

horizonpioneer replied:
컨트롤러에서부터 전달하는 방식에 대해 더 고민해보겠습니다! 감사합니다!

horizonpioneer replied:
혹시 말씀하신 방법이 인자를 통해 LocalDate를 전달하는 방식일까요?
70172f597b8bb10bd154199e415df5ba3c576dcb 여기 타임머신에서 INITIAL_DATE를 LocalDate.now()로 설정해서 시간을 직접 지정하지 않을 경우엔 dateOfNow 정적 메서드를 통해 오늘 날짜를 사용하는 방식은 어떨까요?

sinb57 replied:
LocalDate를 전달하는 방식을 의도했고, 이를 통해 도메인에서 TimeMachine에 대한 의존을 끊길 원했어요. TimeMachine와 같이 static으로 시간을 조작하는 행위는 추후 시간에 얽힌 테스트 작성에 모킹이 불가피 해져요. 되도록 인자로 주고받으며 처리하는 것이 단위테스트 작성시에도 편해요.

horizonpioneer replied:
인자로 주고받는 방식에 대해 고민해봐야겠네요 감사합니다!

sinb57 commented on src/test/java/domain/AttendanceTest.java:
```diff
@@ -0,0 +1,62 @@
+        assertThatThrownBy(() -> new Attendance(localDateTime))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageStartingWith("[ERROR] 주말은 등교일이 아닙니다.");
+    }
```
ParamterizedTest로 중복을 제거할 수 있겠네요. 필요하다면 `Named`를 활용해보세요.

sinb57 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,31 @@
+
+        if (attendanceTime.isAfter(standardTime.getLateTime())) return AttendanceStatus.LATENESS;
+
+        return AttendanceStatus.ATTENDANCE;
```
AttendanceStatus가 아니라 StandardTime가 지녀도 좋지 않을까 싶어요.

sinb57 commented on src/test/java/domain/AttendancesTest.java:
```diff
@@ -0,0 +1,169 @@
+
+class AttendancesTest {
+
+    private static Attendances attendances;
```
테스트 객체를 공유하는 행위는 자칫 테스트 격리에 실패할 수 있으므로 지양해주세요

horizonpioneer replied:
해당 테스트에서 @BeforeEach를 통해 각 테스트에서 독립적으로 적용되도록 했고, 내부 Map.of를 통해 불변객체로 주입되도록 했는데 혹시 다른 문제가 있을까요...?

horizonpioneer replied:
해당 테스트에서 `@BeforeEach`를 통해 각 테스트에서 독립적으로 적용되도록 했고, 내부 Map.of를 통해 불변객체로 주입되도록 했는데 혹시 다른 문제가 있을까요...?

sinb57 replied:
지금은 큰 공감이 되지 않을 수 있지만, 특정 상황에서는 속도 개선 등의 목적으로 parallel 테스트 동작을 세팅할 수 있어요. 이런 경우 할당된 객체가 다른 테스트에서 사용되는 등 예상치 못한 사이드 이펙트가 발생할 수 있다보니 조심해야 해요.

horizonpioneer replied:
단위테스트마다 객체를 공유하기 보단 객체를 생성하는 메서드를 만들어서 각 테스트마다 초기화하도록 하는게 좋겠군요! 공유 객체를 제거해보았습니다

sinb57 commented on src/main/java/domain/Attendances.java:
```diff
@@ -0,0 +1,109 @@
+
+public class Attendances {
+
+    private final Map<String, List<Attendance>> attendances;
```
String은 무엇을 나타내나요? 원시값을 포장해주세요~

sinb57 commented on src/test/java/domain/AttendanceStatusTest.java:
```diff
@@ -0,0 +1,99 @@
+
+        AttendanceStatus attendanceStatus = AttendanceStatus.judge(localDateTime);
+
+        Assertions.assertThat(attendanceStatus.getName()).isEqualTo("출석");
```
attendanceStatus의 name은 입출력에 관련된 필드일텐데, getName으로 꺼내서 비교하는게 옳을까요?

horizonpioneer replied:
Enum 객체 자체로 비교하는게 좋을 것 같네요
처음엔 객체 자체 비교로 테스트했다가 Enum 내부에 각 상수의 필드로 설정한 값을 확인시켜야 한다는 생각에 변경했었습니다.

sinb57 commented on src/test/java/domain/AttendancesLoaderTest.java:
```diff
@@ -0,0 +1,57 @@
+        Files.deleteIfExists(path);
+    }
+
+    @Test
+    void 크루의_출석부를_불러온다() throws IOException {
```
잘못된 포맷의 파일을 읽는 등의 에러 테스트도 작성해주세요

sinb57 commented on src/main/java/domain/Attendance.java:
```diff
@@ -1,45 +1,36 @@
 
-    public Day getDay() {
-        return day;
+    public boolean isEqualTo(LocalDate date) {
+        return localDateTime.toLocalDate().equals(date);
```
테스트 작성해주세요~

sinb57 reviewed (changes requested):
안녕하세요 제프~ 2단계 진행하느라 고생 많으셨어요! 코멘트 남겨두었으니 확인 부탁드려요~

아래의 피드백은 추가로 남깁니다.

- 아직 작성하지 않은 StandardTime 등 모든 도메인 객체의 열려있는 메서드를 추가로 작성해주세요.
- 메서드의 길이 제약 등 프로그래밍 요구 사항을 지켜주세요.

---

> 2. Enum 내부 정적 메서드를 도메인 로직에서 사용하는 것이 적절할까요?

도메인이 Enum에 의존하는 것은 잘못된 방법이 아니지만, 의존 방향에 대해서는 충분한 고민 후 설정한는 것이 중요해요. 어떠한 도메인이 변경이 적으며, 변경에 취약하고 변경의 사이드 이펙트가 강한지 등 충분한 고민을 선행해야만 해요.

> 3. LocalDateTime을 필드로 두는 객체에서 getter를 구현하는 방식이 고민됩니다.
> 출력 시 데이터 가공이 많아 getLocalDate()와 getLocalTime()을 별도로 제공하는 것이 좋을지, 아니면 getLocalDateTime()을 그대로 노출한 뒤 외부에서 가공하는 것이 좋을지 고민됩니다.

무조건 인스턴스 필드를 그대로 반환하는 것이 getter는 아니에요. 때문에 필요에 따라 가공하여 반환해도 문제 없어요.

> 3. LocalDateTime을 필드로 두는 객체에서 getter를 구현하는 방식이 고민됩니다.
> 필드를 LocalDateTime으로 두는 것이 나을지, 아니면 아예 LocalDate와 LocalTime을 분리하는 것이 나을지도 궁금합니다. 상황에 따라 두 방식 모두 사용할 수 있을 것 같지만, 이렇게 하나의 데이터로 관리할 수 있는 필드를 굳이 분리하는 것이 적절한지 조언 부탁드립니다.

상황에 따라 다릅니다! 그치만 가능하면 필요한 정보만을 전달하고 전달받는 것이 중요해요. 코드를 살펴보니 대체로 LocalDateTime으로 전달받고는 필요한 정보를 추출/가공하여 사용하고 있어요. 그래서 본래의 로직이 묻혀서 잘 읽히지 않네요. 이러한 작업은 넘겨주는 쪽에서 처리하는 것이 좋고 메서드의 재사용성이 증가해요. 보통 이런 부분은 TDD 과정에서 메서드 시그니처 등에 대한 피드백을 받으며 개선하곤 해요.

horizonpioneer committed:
feat: 요일에 대한 출석기준시간을 가져오는 기능 테스트 추가

horizonpioneer committed:
feat: 날짜 비교 메서드 테스트 추가

horizonpioneer committed:
feat: 잘못된 포맷을 읽을 경우 예외 발생 테스트 추가

horizonpioneer committed:
refactor: 출석 상태를 판단하는 메서드를 StandardTime 내부로 이동

horizonpioneer committed:
refactor: 주말 테스트 중복 제거

horizonpioneer committed:
refactor: 닉네임 원시값 포장

horizonpioneer committed:
refactor: csv 파일 수정

horizonpioneer committed:
feat: 닉네임 검증 테스트 추가

horizonpioneer committed:
refactor: 타임머신을 통해 오늘 일자를 입력받도록 수정

- 타임머신 테스트 추가
- 어플리케이션 실행 시, 오늘 일자를 입력받도록 설정
- 빈 출석 기록을 결석처리할 때, 오늘 일자까지만 처리되도록 설정

horizonpioneer committed:
feat: 출석 기록에 비어있는 부분을 결석처리 하는 테스트 추가

sinb57 commented on src/test/java/domain/AttendancesLoaderTest.java:
```diff
@@ -0,0 +1,81 @@
+            빙티-2024-12-13 10:07
+            """})
+    void 잘못된_포맷을_읽을_경우_예외를_발생시킨다(String csvFileFormat) throws IOException {
+        BufferedWriter writer = new BufferedWriter(new FileWriter("testAttendance.csv"));
+        writer.write(csvFileFormat);
+        writer.flush();
+        writer.close();
```
중복 로직은 메서드를 분리하여 재사용해보면 어떨까요?

sinb57 commented on src/main/java/domain/AttendancesLoader.java:
```diff
@@ -0,0 +1,49 @@
+
+    public Attendances load(FileReader fileReader) throws IOException {
+        BufferedReader reader = new BufferedReader(fileReader);
+        Map<Nickname, List<Attendance>> attendances = new HashMap<>();
+        try {
+            addAttendanceLogs(reader, attendances);
+        } catch (Exception e) {
+            throw new IOException("[ERROR] 출석 파일을 읽는 중 오류가 발생했습니다.");
+        }
+
+        return new Attendances(attendances);
```
`Map<Nickname, List<Attendance>> attendances` 를 만들어서 값을 모두 생성한 후 `Attendances`의 생성자로 전달하는 구조로 작성하셨네요. 데이터 위주의 코드인데 `Attendances`를 적극 활용해서 객체지향의 코드로 개선해보세요.

---
앞서 남겨주셨던 질문에 대한 답변을 남기기에 적당한 예시인 것 같네요.
(+ 어제 작성해두었는데 코멘트가 안 달렸군요,,)

> 1. 검증의 책임만 가진 객체도 존재 가치가 있을까요?
> ... 검증 책임을 Attendances 객체가 가져야 하는지 고민했습니다. 현재 AttendancesLoader를 통해 출석부 데이터를 불러와 Attendances로 반환하고, Attendances에서 출석부의 key 값에 특정 닉네임이 포함되는지를 검증하는 방식으로 구현했습니다. ... 또 다른 이유는 AttendancesLoader에서 파일을 한 줄씩 읽어 닉네임과 출석 날짜를 파싱하는 과정에서, 닉네임만 따로 모으는 메서드를 추가하면 파일을 두 번 읽어야 하므로 비효율적이라고 판단했기 때문입니다. ...

왜 두 과정을 분리해야 하나요? 한줄씩 읽으면서 출석 정보를 저장하면 안되나요??

> 4. TDD에서 실패 코드 작성이 오래 걸렸습니다.
> ... 도메인에 존재하지 않는 기능을 먼저 테스트 코드에서 가정한 뒤, 해당 기능을 도메인에 추가하고 테스트를 실행한 후 리팩토링하는 방식으로 진행했습니다. ...

지난 1단계 미션에서 테스트 코드가 주는 피드백에 대해 [질문](https://github.com/woowacourse/java-attendance/pull/53#issuecomment-2676855212)드린 적 있어요. 이 때 제프는 코드의 품질을 개선할 수 있다고 말씀주셨는데요. 저 또한 품질 개선 측면에서 큰 장점이 있다고 생각해요. 저의 경우, 테스트 코드를 작성하면서 `외부의 관점으로 사고하는 습관`을 갖고자 했어요. 내부 구현을 작성하는 제공자의 입장이 아니라, 구현을 모른채 외부의 인터페이스(메서드 시그니처)만으로도 의도를 파악하고 적절히 사용할 수 있도록 외부 사용자의 시선으로 바라보고자 노력했어요. 이러니 메서드의 의미 부여 뿐만 아니라 객체의 역할/책임을 명확히 분리할 수 있게 되었고, 자연스레 객체 간의 유기적인 결합을 형성할 수 있었습니다.

제프도 이번 level1에서 단위테스트를 작성하며 많은 부분을 학습하고 스스로의 스타일을 정립해 나갔으면 좋겠어요.

horizonpioneer replied:
1. AttendancesLoader에서 출석정보를 리스트 형태로 반환하여 Attendances를 알지 못하도록 했고, Attendances에 해당 리스트를 인자로 받아 초기화할 수 있도록 하는 메서드를 두어 결합도를 낮추었습니다. 첫 PR에서 닉네임과 출석날짜를 분리해야하는지  질문드렸던 이유는 record나 Dto 형식의 객체를 사용하지 않고 데이터를 전달할 방법이 생각나지 않았기 때문입니다! Pair와 같은 제네릭을 사용한 객체도 생각해봤지만 그것도 record 형식과 비슷하다고 생각했고, 결국 두 가지 데이터를 함께 처리하려면 데이터 운반 객체가 필요하다고 판단해 record 타입의 AttendanceLog를 두었습니다. 

2. 테스트 코드를 작성할 때, 외부 사용자의 시선으로 생각보도록 해야겠습니다! 좋은 인사이트 감사드립니다. 
말씀해주신 테스트 관련 피드백을 바탕으로 더 좋은 테스트 방식이 무엇일지 지속적으로 생각해보겠습니다

(+ 추가적으로 함수명을 자세하게 적는 것에 대해 어떻게 생각하시는지 여쭤보고 싶습니다. 의미가 최대한 전달되도록 짓다보니, 코드가 줄마다 길어지더라구요. 이 부분도 역시 가독성과 의미 전달 사이의 중간점을 찾아야겠죠...?)

sinb57 replied:
> 추가적으로 함수명을 자세하게 적는 것에 대해 어떻게 생각하시는지 여쭤보고 싶습니다. 의미가 최대한 전달되도록 짓다보니, 코드가 줄마다 길어지더라구요. 이 부분도 역시 가독성과 의미 전달 사이의 중간점을 찾아야겠죠...?

의도를 구체적으로 담을수록 내부 구현의 변경에 영향을 받게 돼요. 따라서 적당한 추상화가 필요해요.

예를 들어 `철수는 자전거를 타고 잠실로 간다`라는 상황을 예로 들어 볼게요.
```java
// 상황을 구체적으로 담는다면 아래와 같이 표현할 수 있어요.
철수.rideBikeTo잠실()

// 만약 장소에 대한 변경이 잦은 상황이라면 아래와 같이 표현할 수 있겠죠.
철수.rideBikeTo(잠실)

// 그런데 이동수단에 대해 드러낼 필요가 없다면, 아래와 같이 아예 추상화할 수도 있어요.
철수.goTo(잠실)
// 이 경우, 이동수단을 바꿀 때 내부 구현만 수정하면 됩니다.

// 허나 이동수단을 상황에 따라 바꾸고 싶다면, 외부에서 주입받는 방식도 가능하겠죠
철수.goTo(잠실, 자전거)
철수.goTo(잠실, 자동차)
```

이름을 어떻게 짓는지는 상황에 따라 달라져요. 어떠한 변경점이 예상되고 예정되어 있는지도 꽤나 중요한 요소 중 하나예요.
무엇보다 SOLID 중 "변경되어야 할 이유는 하나이다"와 "변경에는 닫혀있고 확장에는 열려있다"를 충족하는 것이 중요해요.

---

위와 별개로 메서드는 이름만 중요하지 않아요. 메서드 시그니처는 반환 타입과 파라미터 또한 중요한 요소에 해당해요.

`Attendances#findLogWithNameAndDate`를 예로 들어볼게요.

해당 메서드는 `public Attendance findLogWithNameAndDate(Nickname nickname, LocalDate date)`로 표현되고 있어요.
메서드와 파라미터에서 중복된 의미 부여가 존재하고, 파라미터에 메서드명이 의존하는 형태를 띄고 있어요.
그렇다면 `public Attendance findLogWith(Nickname nickname, LocalDate date)`는 어떨까요?
이렇게 한다고 의미가 사라질까요? 오히려 파라미터에 대한 의존을 없앴고, 메서드명은 짧아졌으며, 메서드 오버로딩을 통해 확장에 용이해졌어요.

이와 같이 메서드 시그니처는 사소하지만 심오한 접근을 요구해요. 그리고 이런 접근은 보통 내부 구현에만 몰두하면 놓치기 쉬워요. 때문에 외부 시선에서 바라보는 연습도 꾸준히 해보세요.

horizonpioneer replied:
의미 전달에 문제가 없는 한에서 적당한 추상화를 통해 메서드 시그니처를 설정하는게 중요하겠군요!
findLogWith 예시에 대해 질문드리고 싶은게 있는데 인자로 들어가는 부분에 대해 메서드 이름에서 표현하지 않으면 해당 메서드를 호출해서 사용할 때 어떤 타입을 넣어야 할지 직관적으로 알기 어렵지 않나요? IDE를 통해 단축키를 누르면 바로 볼 수 있기도 하고 해당 메서드를 만든 제 입장에선 바로 알고 쓸 수 있는데, 협업을 진행하는 경우 해당 메서드를 들어가봐야 타입을 알 수 있지 않을까 해서요. 아니면 IDE 기능을 통해 타입을 알 수 있는 부분도 감안해서 이름을 지으면 되는걸까요?

sinb57 replied:
좋은 질문이에요!! 말씀주신 부분도 지극히 맞는 말이지만, 오늘날의 IDE 중 메서드 시그니처를 나열하여 보여주지 않는 IDE는 찾아보기 어려워요. 대부분의 IDE가 개발 생산성을 위해 아래와 같이 메서드를 보여주죠. 이때 파라미터 타입까지도 함께 보이기 때문에, 저는 더더욱 메서드명에 파라미터를 녹일 필요가 없다고 생각해요. 오히려 파라미터의 변경이 메서드명에 영향을 주게 되고, 파라미터를 녹인 메서드명은 오버로딩의 장점을 가져가지 못해요.

<img width="657" alt="image" src="https://github.com/user-attachments/assets/55296e65-26c4-4d19-91c6-ff7b47f41faf" />

sinb57 reviewed (changes requested):
안녕하세요 제프~ 연휴에도 미션 수행하느라 고생 많으셨어요!! 코멘트 확인 부탁드려요~

horizonpioneer committed:
refactor: 테스트 중복 로직 메서드 분리

horizonpioneer committed:
refactor: 공유 테스트 객체 제거

horizonpioneer committed:
refactor: 출석 파일을 로드해서 출석부로 만드는 전체 로직 수정

sinb57 commented on src/test/java/domain/StandardTimeTest.java:
```diff
@@ -0,0 +1,132 @@
+        Assertions.assertThat(attendanceStatus).isEqualTo(AttendanceStatus.ATTENDANCE);
+    }
+
+    @Test
+    void 월요일_출석_경계값_테스트() {
+        LocalDateTime localDateTime = LocalDateTime.of(2024, 12, 16, 13, 5);
+
+        AttendanceStatus attendanceStatus = StandardTime.judge(localDateTime);
+
+        Assertions.assertThat(attendanceStatus).isEqualTo(AttendanceStatus.ATTENDANCE);
+    }
+
+    @Test
+    void 월요일_지각_테스트() {
+        LocalDateTime localDateTime = LocalDateTime.of(2024, 12, 16, 13, 6);
+
+        AttendanceStatus attendanceStatus = StandardTime.judge(localDateTime);
+
+        Assertions.assertThat(attendanceStatus).isEqualTo(AttendanceStatus.LATENESS);
+    }
```
요일별 테스트가 중복 구조를 취하고 있네요. ParameterizedTest를 적용해보면 어떨까요? 필요하다면 Named를 사용해 봐도 좋아요.

horizonpioneer replied:
테스트 명칭 부여 후, 인자로 테스트 값들을 전달해서 확인하는 방식으로 중복을 제거했습니다!

sinb57 reviewed (changes requested):
안녕하세요 제프~ 질문 주신 부분에 답변 남겼어요! 확인 부탁드려요~
2단계 진행도 잘 해주셔서 추가 코멘트는 따로 없어요.
진행 중인 질문 스레드가 종료되면 다음번에 머지할게요

horizonpioneer committed:
refactor: named를 사용하여 테스트 명칭 부여 후 중복 제거

sinb57 reviewed (approved):
안녕하세요 제프~ 1단계에서 구현 및 설계에 대한 얘기를 많이 주고 받은 덕분에, 2단계에서는 테스트 외 따로 리뷰드릴 내용이 없었네요ㅎ 2단계 진행하시느라 고생 많으셨어요!! 출석 미션은 여기서 마무리 하도록 할게요~ 새로 시작하는 블랙잭 미션도 응원할게요! 홧팅입니다!!
